File: scripts/ops/_usage_ledger.py
#!/usr/bin/env python
"""DeepSeek 调用台账（Python 侧）。与 lib/ai/usageLedger.js 写同一个文件、同一种行格式：

  <仓库根>/.ops/deepseek-usage.jsonl   （.ops/ 已被 .gitignore 忽略）
  {"ts","script","label","model","prompt_tokens","completion_tokens","total_tokens",
   "cached_tokens","ms","ok","error","pid"}

环境变量：
  DEEPSEEK_USAGE_LOG     自定义台账路径；设为 0 关闭
  DEEPSEEK_CNY_PER_MTOK  估价用的混合单价（¥/百万 tokens），默认 5.24 = 2026-09-01 账单实测
  TOEFL_ROOT             仓库根覆盖（默认按本文件位置推两级）

record() 永远不抛错——台账写失败不能拖垮正在跑的批量任务。
"""
import os, sys, json, time, datetime

_DEFAULT_RATE = 5.24
_count = 0
_tokens = 0


def repo_root():
    env = os.environ.get("TOEFL_ROOT", "").strip()
    if env:
        return env
    return os.path.abspath(os.path.join(os.path.dirname(os.path.abspath(__file__)), "..", ".."))


def ledger_path():
    v = os.environ.get("DEEPSEEK_USAGE_LOG", "").strip()
    if v and v != "0":
        return v
    return os.path.join(repo_root(), ".ops", "deepseek-usage.jsonl")


def enabled():
    return os.environ.get("DEEPSEEK_USAGE_LOG", "").strip() != "0"


def cny_per_mtok():
    try:
        v = float(os.environ.get("DEEPSEEK_CNY_PER_MTOK", "") or _DEFAULT_RATE)
        return v if v > 0 else _DEFAULT_RATE
    except ValueError:
        return _DEFAULT_RATE


def _int(v):
    try:
        return int(v or 0)
    except (TypeError, ValueError):
        return 0


def default_script_name():
    return os.path.basename(sys.argv[0] or "") or "unknown"


def record(script=None, label=None, model=None, usage=None, ms=None, ok=True, error=None):
    """追加一行；返回 True/False，绝不抛错。"""
    global _count, _tokens
    if not enabled():
        return False
    try:
        usage = usage or {}
        details = usage.get("prompt_tokens_details") or {}
        cached = usage.get("prompt_cache_hit_tokens")
        if cached is None:
            cached = details.get("cached_tokens")
        row = {
            "ts": datetime.datetime.utcnow().strftime("%Y-%m-%dT%H:%M:%S.%f")[:-3] + "Z",
            "script": script or default_script_name(),
            "label": label if label is not None else "",
            "model": model or "",
            "prompt_tokens": _int(usage.get("prompt_tokens")),
            "completion_tokens": _int(usage.get("completion_tokens")),
            "total_tokens": _int(usage.get("total_tokens")),
            "cached_tokens": _int(cached),
            "ms": _int(ms),
            "ok": bool(ok),
            "pid": os.getpid(),
        }
        if error:
            row["error"] = str(error)[:200]
        p = ledger_path()
        d = os.path.dirname(p)
        if d and not os.path.isdir(d):
            os.makedirs(d, exist_ok=True)
        with open(p, "a", encoding="utf-8") as fh:
            fh.write(json.dumps(row, ensure_ascii=False) + "\n")
        _count += 1
        _tokens += row["total_tokens"]
        if _count % 25 == 0:
            print(_summary_line(), file=sys.stderr, flush=True)
        return True
    except Exception:
        return False


def _summary_line():
    return "[deepseek-usage] 本进程已调用 %d 次 · %d tokens · ≈¥%.2f" % (
        _count, _tokens, _tokens * cny_per_mtok() / 1e6)


def session_summary():
    """给脚本结尾打印用；本进程没记录过就返回空串。"""
    return _summary_line() if _count else ""


def recent_stats(script=None, n=50):
    """最近 n 条成功记录的平均 token（同名脚本优先，没有就用全部）。返回 None 或
    {"count","avg_prompt","avg_completion","avg_total","scope"}。给 --dry-run 估价用。"""
    try:
        p = ledger_path()
        if not os.path.exists(p):
            return None
        rows = []
        with open(p, encoding="utf-8") as fh:
            for line in fh:
                line = line.strip()
                if not line:
                    continue
                try:
                    r = json.loads(line)
                except ValueError:
                    continue
                if r.get("ok") and _int(r.get("total_tokens")) > 0:
                    rows.append(r)
        scope = "same-script"
        pick = [r for r in rows if script and r.get("script") == script]
        if not pick:
            pick, scope = rows, "all-scripts"
        pick = pick[-n:]
        if not pick:
            return None
        k = float(len(pick))
        return {
            "count": len(pick),
            "avg_prompt": sum(_int(r.get("prompt_tokens")) for r in pick) / k,
            "avg_completion": sum(_int(r.get("completion_tokens")) for r in pick) / k,
            "avg_total": sum(_int(r.get("total_tokens")) for r in pick) / k,
            "scope": scope,
        }
    except Exception:
        return None

File: lib/ai/usageLedger.js
/**
 * DeepSeek 调用台账 (usage ledger)
 *
 * 背景：2026-09-01 一次本地批量脚本跑掉 518 次 DeepSeek 调用 (¥11.97)，事后只能靠
 * 账单倒推——本地脚本直连 API、没有任何计数。这个模块让每一次调用都在本地留下一行
 * jsonl，配合 scripts/ops/deepseek-usage-report.mjs 做事后对账。
 *
 * 设计约束：
 *  - 模块顶层零副作用（不读文件、不注册监听），否则 jsdom 测试会被污染；
 *  - recordUsage 永不抛错，台账写失败绝不能影响真正的业务调用；
 *  - Vercel / edge runtime 上文件系统不可持久写，直接 no-op。
 */

const fs = require("fs");
const path = require("path");

// 9/1 账单实测的混合单价 (¥/M tokens)：¥11.97 / 2.28M tokens ≈ 5.24
const DEFAULT_CNY_PER_MTOK = 5.24;
// 每记满这么多次在 stderr 报一次进度，防止「跑飞了没人发现」
const PROGRESS_EVERY = 25;
const ERROR_MAX_LEN = 200;

function repoRoot() {
  return path.resolve(__dirname, "../..");
}

/** 台账文件路径：DEEPSEEK_USAGE_LOG 覆盖，否则 <repo>/.ops/deepseek-usage.jsonl */
function resolveLedgerPath() {
  const raw = process.env.DEEPSEEK_USAGE_LOG;
  if (typeof raw === "string") {
    const trimmed = raw.trim();
    if (trimmed && trimmed !== "0") return trimmed;
  }
  return path.join(repoRoot(), ".ops", "deepseek-usage.jsonl");
}

/** DEEPSEEK_USAGE_LOG=0 关闭；Vercel / edge runtime 上强制 no-op */
function isLedgerEnabled() {
  if (process.env.DEEPSEEK_USAGE_LOG === "0") return false;
  if (process.env.VERCEL) return false;
  if (process.env.NEXT_RUNTIME === "edge") return false;
  return true;
}

/** 调用方没传 script 时用入口脚本文件名兜底 */
function defaultScriptName() {
  try {
    return path.basename(process.argv[1] || "") || "unknown";
  } catch (_) {
    return "unknown";
  }
}

/** ¥/M tokens 费率，可用 DEEPSEEK_CNY_PER_MTOK 覆盖 */
function cnyPerMtok() {
  return Number(process.env.DEEPSEEK_CNY_PER_MTOK) || DEFAULT_CNY_PER_MTOK;
}

function toInt(value) {
  const n = Number(value);
  return Number.isFinite(n) ? Math.round(n) : 0;
}

function estimateCny(totalTokens, rate = cnyPerMtok()) {
  return (Number(totalTokens) || 0) * (Number(rate) || 0) / 1e6;
}

/** 小额用 4 位小数，避免真实成本被四舍五入成 ¥0.00 */
function formatCny(value) {
  const n = Number(value) || 0;
  return n >= 1 ? n.toFixed(2) : n.toFixed(4);
}

// 进程内累计计数。惰性创建 —— 顶层不能有副作用。
let counters = null;

function ensureCounters() {
  if (!counters) counters = { calls: 0, tokens: 0, exitHooked: false };
  return counters;
}

function progressLine(c) {
  return `[deepseek-usage] 本进程已调用 ${c.calls} 次 · ${c.tokens} tokens · ≈¥${formatCny(estimateCny(c.tokens))}`;
}

function emitProgress(c) {
  if (process.env.DEEPSEEK_USAGE_QUIET === "1") return;
  try {
    process.stderr.write(`${progressLine(c)}\n`);
  } catch (_) {
    /* stderr 不可写也不能炸 */
  }
}

function ensureExitHook(c) {
  if (c.exitHooked) return;
  c.exitHooked = true;
  try {
    process.on("exit", () => {
      if (c.calls >= 1) emitProgress(c);
    });
  } catch (_) {
    /* 注册失败就算了 */
  }
}

/**
 * 追加一行台账。永不抛错；禁用或写失败时返回 false。
 *
 * @param {object} entry
 * @param {string} [entry.script] 调用方脚本名（默认取 argv[1] 的 basename）
 * @param {string} [entry.label]  自定义标签（题型 / 阶段等）
 * @param {string} [entry.model]  DeepSeek model
 * @param {object} [entry.usage]  DeepSeek 原样的 data.usage（缺字段容错为 0）
 * @param {number} [entry.ms]     耗时
 * @param {boolean} [entry.ok]    是否成功
 * @param {string} [entry.error]  失败原因（截 200 字）
 * @returns {boolean} 是否真的写进去了
 */
function recordUsage(entry) {
  try {
    if (!isLedgerEnabled()) return false;

    const e = entry || {};
    const usage = e.usage || {};
    const promptTokens = toInt(usage.prompt_tokens);
    const completionTokens = toInt(usage.completion_tokens);
    const totalTokens = toInt(usage.total_tokens) || promptTokens + completionTokens;
    const cachedTokens = toInt(
      usage.prompt_cache_hit_tokens ?? usage.prompt_tokens_details?.cached_tokens ?? 0,
    );

    const row = { ts: new Date().toISOString(), script: String(e.script || defaultScriptName()) };
    if (e.label) row.label = String(e.label);
    row.model = String(e.model || "unknown");
    row.prompt_tokens = promptTokens;
    row.completion_tokens = completionTokens;
    row.total_tokens = totalTokens;
    row.cached_tokens = cachedTokens;
    row.ms = toInt(e.ms);
    row.ok = e.ok !== false;
    if (e.error) row.error = String(e.error).slice(0, ERROR_MAX_LEN);
    row.pid = process.pid;

    const filePath = resolveLedgerPath();
    const dir = path.dirname(filePath);
    if (dir && !fs.existsSync(dir)) fs.mkdirSync(dir, { recursive: true });
    fs.appendFileSync(filePath, `${JSON.stringify(row)}\n`, "utf8");

    const c = ensureCounters();
    ensureExitHook(c);
    c.calls += 1;
    c.tokens += totalTokens;
    if (c.calls % PROGRESS_EVERY === 0) emitProgress(c);

    return true;
  } catch (_) {
    // 台账是观测设施，永远不许影响主链路
    return false;
  }
}

/** 读台账为对象数组（坏行跳过；文件不存在返回 []） */
function readLedger(filePath) {
  const target = filePath || resolveLedgerPath();
  let raw = "";
  try {
    raw = fs.readFileSync(target, "utf8");
  } catch (_) {
    return [];
  }
  const rows = [];
  for (const line of String(raw).split(/\r?\n/)) {
    const text = line.trim();
    if (!text) continue;
    try {
      const parsed = JSON.parse(text);
      if (parsed && typeof parsed === "object") rows.push(parsed);
    } catch (_) {
      /* 坏行跳过 */
    }
  }
  return rows;
}

module.exports = {
  DEFAULT_CNY_PER_MTOK,
  resolveLedgerPath,
  isLedgerEnabled,
  defaultScriptName,
  cnyPerMtok,
  estimateCny,
  formatCny,
  recordUsage,
  readLedger,
};
